research/lamp-automated-model-parallelism/train.py

```diff
@@ -0,0 +1,212 @@
+from torchgpipe import GPipe
+from torchgpipe.balance import balance_by_size
+
+from unet_pipe import UNet, flatten_sequential
+from data_utils import get_filenames, load_data_and_mask
+
+N_CLASSES = 10
+TRAIN_PATH = "./data/HaN/train/"  # training data folder
+VAL_PATH = "./data/HaN/test/"  # validation data folder
+
+torch.backends.cudnn.enabled = True
+
+
+class ImageLabelDataset:
+    """
+    Load image and multi-class labels based on the predefined folder structure.
+    """
+
+    def __init__(self, path, n_class=10):
+        self.path = path
+        self.data = sorted(os.listdir(path))
+        self.n_class = n_class
+
+    def __getitem__(self, index):
+        data = os.path.join(self.path, self.data[index])
+        train_data, train_masks_data = get_filenames(data)
+        data = load_data_and_mask(train_data, train_masks_data)  # read into a data dict
+        # loading image
+        data["image"] = data["image"].astype(np.float32)  # shape (H W D)
+        # loading labels
+        class_shape = (1,) + data["image"].shape
+        mask0 = np.zeros(class_shape)
+        mask_list = []
+        flagvect = np.ones((self.n_class,), np.float32)
+        for i, mask in enumerate(data["label"]):
+            if mask is None:
+                mask = np.zeros(class_shape)
+                flagvect[0] = 0
+                flagvect[i + 1] = 0
+            mask0 = np.logical_or(mask0, mask)
+            mask_list.append(mask.reshape(class_shape))
+        mask0 = 1 - mask0
+        data["label"] = np.concatenate([mask0] + mask_list, axis=0).astype(np.uint8)  # shape (C H W D)
+        # setting flags
+        data["with_complete_groundtruth"] = flagvect  # flagvec is a boolean indicator for complete annotation
+        return data
+
+    def __len__(self):
+        return len(self.data)
+
+
+def train(n_feat, crop_size, bs, ep, optimizer="rmsprop", lr=5e-4, pretrain=None):
+    crop_size = [int(cz) for cz in crop_size.split(",")]
+    print(f"input image crop_size: {crop_size}")
+
+    # starting training set loader
+    train_images = ImageLabelDataset(path=TRAIN_PATH, n_class=N_CLASSES)
+    if np.any([cz == -1 for cz in crop_size]):  # using full image
+        train_transform = Compose([AddChannelDict(keys="image")])
+        train_dataset = Dataset(train_images, transform=train_transform)
+        # when bs > 1, the loader assumes that the full image sizes are the same across the dataset
+        train_dataloader = torch.utils.data.DataLoader(train_dataset, num_workers=6, batch_size=bs, shuffle=True)
+    else:
+        train_transform = Compose(
+            [
+                AddChannelDict(keys="image"),
+                CopyItemsd(keys="label", times=1, names="label_fg"),
+                lambda d: d.update({"label_fg": d["label_fg"][1:]}) or d,  # excluding label bg from the sampling mask
+                RandCropByPosNegLabeld(keys=("image", "label"), label_key="label_fg", size=crop_size, num_samples=bs),
```

Comment:
looks like `RandCropByPosNegLabeld`'s `label_key` doesn't accept one-hot labels that have both foreground and background channels? here I use a copyitem to implement a workaround, any comments/suggestions @Nic-Ma?

Reply:
Hi @wyli ,

Thanks for our bug report!
I made a quick update: https://github.com/Project-MONAI/MONAI/pull/620, could you please help review it?

Thanks.

Reply:
> Hi @wyli ,
> 
> Thanks for our bug report!
> I made a quick fix: #620, could you please help review it?
> 
> Thanks.

thanks it works